includes/Classifai/Providers/OpenAI/ChatGPT.php

```diff
@@ -0,0 +1,510 @@
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Register what we need for the plugin.
+	 *
+	 * This only fires if can_register returns true.
+	 */
+	public function register() {
+		add_action( 'enqueue_block_editor_assets', [ $this, 'enqueue_editor_assets' ] );
+	}
+
+	/**
+	 * Enqueue the editor scripts.
+	 */
+	public function enqueue_editor_assets() {
+		$settings = $this->get_settings();
+
+		// Don't load our custom post excerpt if excerpt functionality isn't turned on.
+		if ( ! isset( $settings['enable_excerpt'] ) || 1 !== (int) $settings['enable_excerpt'] ) {
+			return;
+		}
+
+		// This script removes the core excerpt panel and replaces it with our own.
+		wp_enqueue_script(
+			'classifai-post-excerpt',
+			CLASSIFAI_PLUGIN_URL . 'dist/post-excerpt.js',
+			get_asset_info( 'post-excerpt', 'dependencies' ),
+			get_asset_info( 'post-excerpt', 'version' ),
+			true
+		);
+	}
+
+	/**
+	 * Setup fields
+	 */
+	public function setup_fields_sections() {
+		$default_settings = $this->get_default_settings();
+
+		// Add the settings section.
+		add_settings_section(
+			$this->get_option_name(),
+			$this->provider_service_name,
+			function() {
+				printf(
+					wp_kses(
+						/* translators: %1$s is replaced with the OpenAI sign up URL */
+						__( 'Don\'t have an OpenAI account yet? <a title="Sign up for an OpenAI account" href="%1$s">Sign up for one</a> in order to get your API key.', 'classifai' ),
+						[
+							'a' => [
+								'href'  => [],
+								'title' => [],
+							],
+						]
+					),
+					esc_url( 'https://platform.openai.com/signup' )
+				);
+			},
+			$this->get_option_name()
+		);
+
+		// Add all our settings.
+		add_settings_field(
+			'api-key',
+			esc_html__( 'API Key', 'classifai' ),
+			[ $this, 'render_input' ],
+			$this->get_option_name(),
+			$this->get_option_name(),
+			[
+				'label_for'     => 'api_key',
+				'input_type'    => 'password',
+				'default_value' => $default_settings['api_key'],
+			]
+		);
+
+		add_settings_field(
+			'enable-excerpt',
+			esc_html__( 'Generate excerpt', 'classifai' ),
+			[ $this, 'render_input' ],
+			$this->get_option_name(),
+			$this->get_option_name(),
+			[
+				'label_for'     => 'enable_excerpt',
+				'input_type'    => 'checkbox',
+				'default_value' => $default_settings['enable_excerpt'],
+				'description'   => __( 'A button will be added to the excerpt panel that can be used to generate an excerpt.', 'classifai' ),
+			]
+		);
+
+		add_settings_field(
+			'length',
+			esc_html__( 'Excerpt length', 'classifai' ),
+			[ $this, 'render_input' ],
+			$this->get_option_name(),
+			$this->get_option_name(),
+			[
+				'label_for'     => 'length',
+				'input_type'    => 'number',
+				'max'           => 5,
+				'min'           => 1,
+				'step'          => 1,
+				'default_value' => $default_settings['length'],
+				'description'   => __( 'How many sentences should the excerpt be?', 'classifai' ),
```

Comment:
Is there any opportunity to limit this to characters instead of sentences with OpenAI or is their prompt the number of sentences only?  I ask as some folks' front end designs might only support up to a certain number of characters for rendering an excerpt and am curious if we would be able to better support that sort of precision versus a random total character count based on number of sentences?

Reply:
We have full control over the prompt we send. Right now that looks like:
`Summarize the following text into 2 short sentences: {post_content}`

There isn't an option on the API side to limit by characters, words or sentences. There is a `max_tokens` param we can use to limit how many tokens the response contains but there's not a super clean way to convert words or characters to tokens (it's roughly 4 characters per token).

I'm open to the idea of changing this, as the responses I usually get from the API are fairly long. I guess I'm concerned that most people won't have a clue on how many characters they want. Maybe we change this from sentences to words? WordPress uses 55 words per excerpt as a default with a filter around that value. Thinking we can pull that same filtered value to use as the default here.

Reply:
All right, I've changed this from setting the excerpt length in sentences to words, defaulting to the excerpt length value in WordPress. Note though that in testing, the results returned from the API tend to exceed this number. It was fairly consistently 40-50 words more than requested and after experimenting with the prompt we send, it's now more consistently around 10-15 extra words.

I've updated the description on the setting to note this but I guess we need to decide if being able to fine-tune a little better by using words is worth the frustration when the results returned will more than likely exceed that number. When requesting sentences, it always gave me the amount of sentences I requested but the results were usually longer than I would want for an excerpt. Now I get results closer to what I would expect for an excerpt but still longer than what I requested.

Reply:
I'm not strongly opinionated in either direction here, so if there's an approach that "feels right" to you then go for it.

Reply:
This feels better to me, we can definitely tweak/change this in the future if we want/get feedback